File: src/training/skill.ts
import { SKILL, SKILL_LABEL, SKILL_POLL, SKILL_TIMEOUT } from './config.js';

// Every read of getSkill goes through here, because the client's own typings declare it returning
// undefined and the whole script turns on what that is taken to mean.

// The client's tenths as a person reads them: 746 is 74.6
export const tenths = (value: number): string => (value / 10).toFixed(1);

// Never coerced to 0. 0 is a real skill value, and reading a blind client as 0 makes the trainer cast
// the first band's ability at a character who has capped the skill - so the maybe is passed on and
// every caller is made to say what it does about it.
export const skillValue = (): number | undefined => player.getSkill(SKILL)?.value;

export const skillCap = (): number | undefined => player.getSkill(SKILL)?.cap;

// The shard's own name for it once the skill list has arrived, and the configured label until then
export const skillName = (): string => player.getSkill(SKILL)?.name ?? SKILL_LABEL;

// The skill list arrives asynchronously like everything else, so a run that opened by reading it once
// would decide what to train from whatever the client happened to know at paste time. Polled, and the
// common case - a list that arrived long ago - costs one read.
export const waitForSkill = (): number | undefined => {
  for (let waited = 0; waited < SKILL_TIMEOUT; waited += SKILL_POLL) {
    const value = skillValue();

    if (value !== undefined) {
      return value;
    }

    sleep(SKILL_POLL);
  }

  return skillValue();
};

File: legion/src/uo/skill.py
import API


def skill_value(name):
    skill = API.GetSkill(name)

    return skill.Value if skill is not None else 0.0


# The client answers 0 until the server has sent the skill list, which is not the same as a skill
# that is genuinely at 0
def wait_for_skill(name, timeout, poll):
    waited = 0.0

    while waited < timeout:
        skill = API.GetSkill(name)

        if skill is not None and skill.Value > 0:
            return skill

        API.Pause(poll)
        waited += poll

    return API.GetSkill(name)
